feat(kubectl): adding --node-name option to top

joshkurz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

Adds an option to kubectl top pod, which prints out hostIP information alongside pod information:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #97859

**Special notes for your reviewer**:
Here is an example of how this could be done. At first I wanted to get the extra pod information from the metrics api, but that was a little more involved than just making 1 extra call to get the pods if the option is set. Overall this does what I intended. It could be a little better. we could have it work with the sort option if we wanted. Could also allow for other pod status data if we wanted as well. But overall hostIP was the most important field that was needed when topping pods.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
## print host ip information alongside pod data when running top
kubectl top pods --wide

```

Commit:
feat(kubectl): adding --node-name option to top

joshkurz commented:
/kind feature

@eddiezane here is an example of what I was thinking about adding.

rikatz commented:
/ok-to-test

rikatz commented:
/assign

rikatz commented:
/cc

joshkurz commented:
/retest

rikatz commented:
@joshkurz Thanks for the PR. I'll start to review it.

Right now:

* Any reason to print Host IP and not the node name? I guess the host name is better suited in this case, the same way when you do a kubectl get pods -o wide and it returns the node name and not the IP Address.
* Can we consider using -o wide instead of --wide? Was this flag discussed? I guess -o is a well known flag that can be used to return the wide option, instead of --wide, wdyt?

Thanks! :)

joshkurz commented:
Yep both of those make sense.

-o wide should print hostName instead of hostIP. I can do that for sure.

One question I have. Should -o wide actually just print all possible options? and for this hostName option we add a new option --hostname? That way a user can just do -o wide and all options get added or they can do them individually like today.

joshkurz commented:
also if we add a -o option, should we support json as well? and other forms of output?

I dont want to take away from the real focus of this PR which is to add the ability to print hostName with top. That is the real feature I think adds the most value. If we want to do -o wide, im game for it. but do you think we should do it in a different PR? And just change this option to --hostname or --host-name?

joshkurz commented:
ok just updated this PR to have --node-name as the option flag. if we want to do -o wide or something like that, might be good to do in another PR. I can see a lot of advantages for it, but not sure if anyone is asking for it. Plus it could introduce a lot more possible code if its to be done right, or some refactoring to reuse a lot of what `get` uses.

Let me know what you think. Open to any suggestions to help out.

rikatz commented:
OK, so another question: per @eddiezane comment on issue #97859  the SIG required a quick presentation about this, so we can broader the discussion. 

Can you please present this? (or had you already did the presentation?)

So we can close the decision of -o wide vs --wide, if it should show now node name or the ip, etc ;)

Next meeting is tomorrow 14:00 (GMT-3, don't know what's your timezone ;))

If you can, please add this to the agenda: https://docs.google.com/document/d/1r0YElcXt6G5mOWxwZiXgGu_X6he3F--wKwg-9UBc29I/edit

joshkurz commented:
Cool, I didn't know y'all had these meetings for demos and things. I'm in EST so that works for me. I added the topic to tomorrows meeting block.

dougsland commented:
/cc

joshkurz commented:
ok so, I agree that it makes sense to do this with -o wide option.

My questions is do we want to add more information than just NodeName if -o wide is present? ie, we can add hostIp as well like get -o wide does.

example would be two extra columns like this: vs just Node. This matches what get -o wide does as well.
```
IP               NODE 
xx.xx.xxx.xxx    ip-xxx-xxx-xxx-xxx.ec2.internal
xx.xx.xxx.xxx    ip-xxx-xxx.xxx.xxx.ec2.internal
```

rikatz commented:
@joshkurz so because we need to make this opt-in (you MUST pass a flag) because of some potential RBAC conflict, maybe --show-nodes is a good choice IMO! :)

Commit:
fix(staging): add back -o wide and HostIP

joshkurz commented:
ok I just repushed it. I updated this time to be 

```
kubectl top pod -o wide
```

If -o wide is not present, this does the normal thing it does today. If it is present, it will print NodeName and HostIP which is what kubectl get -o wide does. Also, if -o wider or -o foobar is passed in, this does nothing, as we are only looking for -o wide.

rikatz commented:
> ok I just repushed it. I updated this time to be
> 
> ```
> kubectl top pod -o wide
> ```
> 
> If -o wide is not present, this does the normal thing it does today. If it is present, it will print NodeName and HostIP which is what kubectl get -o wide does. Also, if -o wider or -o foobar is passed in, this does nothing, as we are only looking for -o wide.

my point of using --show-nodes, --show-details (or other unknown flag) instead of -o wide, is that -o wide is a well known flag used in other commands that can lead to errors to users (because of the RBAC problem explained before), while --show-nodes is a new unknown flag that the used needs to know that it exists and we can put somewhere that "to use this flag, user must have permission to fetch metrics and also get Pods" :)

About the IP, I would complement the column name to be "HOST IP" instead of only IP, to not make confusion with Pod IP (which IMO, can be also part of that flag, now that we already make the additional API call)

But this is my opinion, will wait also for some other reviewers :D

joshkurz commented:
@rikatz my understanding is, since -o isnt a current option on `top`, this isn't a breaking change. 
Also its better that we follow the same type of conventions that `get` uses, since more people are familiar with that output and option. 

I'm ok with whatever we decide, this was just I took away from the meeting we had. That we thought a -o wide was a better option than a --node-name or something that does not follow the same convention as `get`.

rikatz commented:
OK, agree with you :)

Just please make sure:

1) To rename the column to Node IP (otherwise this can lead to some users confusion if this is the Pod or the Node IP)
2) To put in the description, or in some docs that using '-o wide' may require additional permissions (I know this is a very corner case, but we need to make sure users hitting 'permission forbidden' when using kubectl top knows why!)
3) To rebase :)

Ping me when you're ready, sorry for the long delay :/

Commit:
fix(kubectl): update column name to NODE_IP

joshkurz commented:
@rikatz thank you for the review. I have rebased and updated the column name to be HOST_IP. 

Where do you want me to update the docs to explain the possible permissions issue?

Commit:
fix(metrics_printer): print host and host_ip for single container option

Commit:
docs(kubectl): adding rbac info to -o wide option print out

joshkurz commented:
@rikatz I updated the description of the option to include information about requiring GET /pods access.

joshkurz commented:
@rikatz I rebased this up again. Should be good for review when you are ready. Thanks.

rikatz commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -200,49 +202,70 @@ func (printer *TopCmdPrinter) PrintPodMetrics(metrics []metricsapi.PodMetrics, p
 		if withNamespace {
 			printValue(w, NamespaceColumn)
 		}
+		if len(pods.Items) > 0 {
```
```suggestion
		if pods != nil && len(pods.Items) > 0 {
```

Otherwise if the pointer is nil it might get a panic :)

joshkurz replied:
would this be possible, since this has to be passed into the function to begin with? if pods is nil, that would mean someone passed in nil directly here correct? because they will get an error if they dont pass pods in.

Also if we do this, should we also add something inside of the createPodMap function?

```
func createPodMap(pods *v1.PodList) (podMap map[string]v1.Pod) {
	podMap = make(map[string]v1.Pod)
	for _, pod := range pods.Items {
		podMap[pod.Name] = pod
	}
	return podMap
}
```

That way we dont have to check explicity for pods when calling this function on line 216.
```
podMap := createPodMap(pods)
```

The thought is, we always call this so we dont have to check if pods exist or not. We only need to check if individual pods exist inside the list.

The default state of pods is ```var pods *v1.PodList = new(v1.PodList)``` so if we get an error when getting pods, we return error and this function fails to begin with. we don't continue.

joshkurz replied:
Trying to keep the code a little clean and not have to check stuff all over the place. This is a default param, so not sure how it would ever be ok for someone to pass nil as the podsList interface is what is expected.

rikatz commented on staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go:
```diff
@@ -200,49 +202,70 @@ func (printer *TopCmdPrinter) PrintPodMetrics(metrics []metricsapi.PodMetrics, p
 	if withNamespace {
 		printValue(out, m.Namespace)
 	}
+	if podExists {
```
hum, can't this be something like "if pod != nil"? Instead of passing the value and the boolean?
Example:
```golang
        podMap := make(map[string]interface{})
	podMap["a"] = "aaa"
	podMap["b"] = "bbb"
	v, _ := podMap["c"]
	if v != nil {
		fmt.Println(v) // Will print nothing
	}
        x, _ := podMap["a"]
	if x != nil {
		fmt.Println(x) // Will print "aaa"
	}
```

joshkurz replied:
Yeah was not 100% sure of the best way to determine if a pod exists in the map. But it looks like the zero value of a interface is nil, so this works.

joshkurz replied:
So actually doing it like this breaks. Since this is mapping to a struct Pod type. the zero value is not nil and this does not work. Its throwing an error. do you have another suggested method or should we continue to use the boolean returned from the map when a key does not exist?

The purpose of this podMap is to make sure we can quickly lookup pods and print out the information relative the pod without having to loop through the PodList every time.

rikatz reviewed (commented):
Hey @joshkurz we are almost there :D

Left some other reviews, and I guess in the next pass we should be good to go! :D

Commit:
fix(kubectl): addressing fixes

Commit:
kubectl(staging): revert back how we test existence of keys

Commit:
fix(kubectl): remove vscode stuff

fedebongio commented:
/remove-sig api-machinery
